### ISSUE TITLE
Custom dataset making and training.

### ISSUE THREAD

**Issue by M1keStr0ng** [facebookresearch/SlowFast#581]
Dear all, I am a college student doing my final year project. I want to train my custom dataset, although there are some tutorials for making dataset, I never success. I know there are some platforms can train slowfast, such as Pyslowfast/Detectron2/mmaction2... and know types of action dataset (kinetics/AVA/UCF...). Is there any kind person can give me some hint about making custom dataset (any kind of dataset is fine) and training? If you can, I will be very appreciate!

**Comment by jhonmajinerazo** [comment]
> Dear all, I am a college student doing my final year project. I want to train my custom dataset, although there are some tutorials for making dataset, I never success. I know there are some platforms can train slowfast, such as Pyslowfast/Detectron2/mmaction2... and know types of action dataset (kinetics/AVA/UCF...). Is there any kind person can give me some hint about making custom dataset (any kind of dataset is fine) and training? If you can, I will be very appreciate!

hello, I also want to train slowfast on custom dataset based on AVA. but I can't find any information. how do i generate the .csv files

**Comment by alpargun** [comment]
Please follow this explanation. https://github.com/facebookresearch/SlowFast/issues/149#issuecomment-723265461

I managed to train on BDD (Berkeley Deep Drive) dataset, which is not included in the SlowFast framework, following this guide. About the train.csv, val.csv, test.csv files, you need to create them on your own.

**Comment by poincarelee** [comment]
how is  the training results?  I trained on HMDB51, the final accuracy is quite low(46% for top1, 72% for top5). Training on K400 seems more weird, final accuracy is 16% for top1, 31% for top5.

**Comment by M1keStr0ng** [comment]
> how is the training results? I trained on HMDB51, the final accuracy is quite low(46% for top1, 72% for top5). Training on K400 seems more weird, final accuracy is 16% for top1, 31% for top5.

Hi, can you make custom dataset?

**Comment by poincarelee** [comment]
alpargun  have mentioned how to make custom dataset, you can refer to the link he supported.

**Comment by alpargun** [comment]
I am running the Self Supervised Learning (SSL) task with SlowFast, hence, I'm not using labels in my data and I won't be able to help you with the comparison of the results. However, with almost half of the BDD dataset, I am seeing the results for SSL training:
- top1 error: 35%
- top5 error: 20% approximately. I am emphasizing that these are the error percentages. So, the accuracy will be (100-error)%
- Also I have a continuously decreasing train/loss graph. As I haven't found enough documentation about how to evaluate SSL training, I am also not 100% sure how to assess my results until I use my trained model on the other task I aim to achieve.

**Comment by poincarelee** [comment]
Thanks.  Your results look good .
I was wondering if you have examined the loss trend or top1 error between different batches, I often found that during one epoch,  some batch's loss(top1-err) was quite different from other batch's.   When averaging all the batches during one epoch,  the loss decreased.

**Comment by alpargun** [comment]
Yes, the errors fluctuate between the iterations and I would suggest using Tensorboard (you need to turn it on in the config file as it isn't enabled by default) to visualize the performance metrics better.

**Comment by poincarelee** [comment]
Good advice, you are right.
I plotted top1 error(top5 err) by matplotlib.  I see in the training code there's option for Tensorboard, I will try it.

**Comment by poincarelee** [comment]
Hi, @alpargun,
I have tried tensorboard, the result is as follows:
![image](https://user-images.githubusercontent.com/38481402/195023482-9c488ccc-adee-43f4-b589-11ebb78efffd.png)
![image](https://user-images.githubusercontent.com/38481402/195023463-66e237cb-0313-4647-8797-e1149d3fff37.png)
the training error is decreasing, while the val loss is fluctuating weirdly. 
The model is training on K400.
Have you got any clue about this? Thanks

**Comment by alpargun** [comment]
The validation results are also strange for me (e.g. sometimes the loss just stays 0). So I am not sure about the validation metrics. However, my training top-k errors and the loss graphs have a similar decreasing trend. More explanation about the evaluation metrics for training on a custom dataset by the SlowFast team would be great.

**Comment by ra1995** [comment]
@alpargun Hi, I saw your recent comments and gather that you have successfully implemented the SSL module for your own data. I am trying to fine tune the weights that are provided in the model zoo for MoCo, by training on my custom unlabelled video data. However the projection MLP head weights are missing. Is this going to an issue while training? Do I need need to make any modifications to the projection head?

**Comment by alpargun** [comment]
Hi @ra1995 
I experienced the same issue where the MLP weights were lost but my training results turned out to be well. So, in practice I did not have any problems with that.
Also when I read the related SSL papers, I see that after finishing an SSL training with unlabeled data, the authors are mostly evaluate their model by fine-tuning or linear-probing, where they remove the last layer(s) for the projection head, and add new MLP head to specialize in new data. Hence, also in theory I don't think not having the MLP head weights loaded will cause any problems after a long enough training.
